ambry-cloud/src/main/java/com/github/ambry/cloud/BackupIntegrityMonitor.java

```diff
@@ -0,0 +1,192 @@
+        .filter(d -> d.getState() == HardwareState.AVAILABLE)
+        .filter(d -> d.getAvailableSpaceInBytes() >= maxReplicaSize)
+        .collect(Collectors.toList());
+    // Pick disk randomly
+    DiskId disk = disks.get(new Random().nextInt(disks.size()));
+    // Convert Disk object to AmbryDisk object, static-map -> helix-map
+    AmbryDisk ambryDisk = new AmbryDisk((Disk) disk, clusterMapConfig);
+    AmbryServerReplica localReplica = new AmbryServerReplica(clusterMapConfig, partition, ambryDisk,
+        true, maxReplicaSize, ReplicaSealStatus.NOT_SEALED);
+    if (!(storageManager.addBlobStore(localReplica) && storageManager.startBlobStore(partition))) {
+      throw new RuntimeException(String.format("Failed to add/start Store for %s", partition.getId()));
+    }
+    Store store = storageManager.getStore(partition, true);
+    logger.info("[BackupIntegrityMonitor] Started Store {}", store.toString());
+    return (BlobStore) store;
+  }
+
+  private boolean stopLocalStore(AmbryPartition partition) {
+    Store store = storageManager.getStore(partition, true);
+    try {
+      boolean ret = storageManager.shutdownBlobStore(partition) && storageManager.removeBlobStore(partition);
+      logger.info("[BackupIntegrityMonitor] Stopped Store [{}]", store);
+      return ret;
+    } catch (Throwable e) {
+      logger.error(String.format("Failed to stop/remove Store [%s]", store));
+      return false;
+    }
+  }
+
+  /**
+   * This method randomly picks a partition in a cluster and locates its replica in the cluster and the cloud.
+   * Then it downloads and compares data from both replicas.
+   */
+  @Override
+  public void run() {
+    RemoteReplicaInfo cloudReplica = null;
+    AmbryPartition partition = null;
+    try {
+      // 1. Pick a partition P, replica R from helix cluster-map
+      // 2. Pick a disk D using static cluster-map
+      // 3. while(!done) { RecoveryThread::replicate(P) } - restore partition P from cloud to disk D
+      // 4. while(!done) { BackupCheckerThread::replicate(R) } - copy metadata from R, compare with metadata in D
+
+      /** Select partition P */
+      // TODO: Don't pick the same partition twice
+      List<PartitionId> partitions = helixClusterManager.getAllPartitionIds(null);
+      partition = (AmbryPartition) partitions.get(new Random().nextInt(partitions.size()));
+      logger.info("[BackupIntegrityMonitor] Verifying backup partition-{}", partition.getId());
+
+      /** Create local Store S */
+      BlobStore store = startLocalStore(partition);
+
+      /** Restore cloud backup C */
+      logger.info("[BackupIntegrityMonitor] Restoring backup partition-{} to disk [{}]", partition.getId(), store);
+      cloudReplica = azureReplicationManager.getCloudReplica(store.getReplicaId());
+      // TODO: Reload tokens
+      azureReplicator.addRemoteReplicaInfo(cloudReplica);
+      while (!((RecoveryToken) cloudReplica.getToken()).isEndOfPartition()) { azureReplicator.replicate(); }
```

Comment:
is there anyway we can print out some message to show progress?

Reply:
yes done